pkg/lib/cockpit-components-typeahead-select.tsx

```diff
@@ -421,7 +430,10 @@ export const TypeaheadSelectBase: React.FunctionComponent<TypeaheadSelectProps>
                     const { content, value, ...props } = option;
 
                     return (
-                        <SelectOption key={value} value={value} isFocused={focusedItemIndex === index} {...props}>
+                        <SelectOption key={index.toString()}
```

Comment:
keys *must not* be something extrinsic that can change, they must be derived from the intrinsic value. This is fundamentally broken, and in such cases there's no other way than letting the caller submit a proper key (or we should just fix pkg/networkmanager to not submit multiple values in the first place -- or disambiguate them in some other way.

Reply:
Well...

Here is my take: Keys matter when there is state involved. Otherwise they are a nuisance without benefit. Red tape. I am happy to shut React up with list indices, and I am even happier when a component does it for me. I am always tempted to construct a React.Fragment at run-time like we do in `fmt_to_fragments`. (And I did that just to shut up React, did you know? :)

Why do we need keys with this

```
const names = [];
if (cond1)
  names.push("A");
if (cond2)
  names.push("B");
<div>
  { names.map(x => <span>{c}</span> }
</div>
```

but not with this

```
<div>
  { cond1 && <span>A</span> }
  { cond2 && <span>B</span> }
</div>
```

Can we even use keys with the latter? Should it be illegal to optionally render things? Maybe it is ok to optionally render things without state. Maybe it is ok to use list indices as keys when there is no state.

The alternative to assigning keys here would be to assign them in the networkmanager code. And what would we do there? Use list indices like in a million other places.

And look at how the old code did things with the deprecated Select:
```
optionsMap.map((option, index) => (
                <SelectOption key={index}
                              value={option}
                />
            )}
```

List indices are fine, they have always been. (When there is no state involve. And the day SelectOption involves state I'll rename PatternFly to MudderFlucker).

</rant> I hope you enjoyed. :-)

Reply:
> The alternative to assigning keys here would be to assign them in the networkmanager code. And what would we do there?

The general solution: Construct a key which includes more data of whatever the item represents. Specific solution: I think it's reasonable to assume that a Select has unique `value`s (i.e. `optionsMap`), and use the `value`s as keys. I.e. do that, and fix that bug on the NM page that duplicates items -- it confuses users as well as React.

> Use list indices like in a million other places.

"A million" is hopefully a big exaggeration. I've spent hours debugging and fixing bugs due to duplicate keys, and I don't want them introduced into a shared component.

> List indices are fine, they have always been

No, they are **not** fine as soon as the entries can change. And we can't make that assumption in a shared component. The old code was indeed just simply broken. Consider it like this: If list indexes *were* fine as keys, then React wouldn't need to have a concept of `key` in the first place. They are being used as optimizing rendering after state changes, and they key determines which ones changed and which ones in the DOM can stay. You can't do that with extrinsic information like list indexes.

Reply:
> I've spent hours debugging and fixing bugs due to duplicate keys, and I don't want them introduced into a shared component.

Using list indices guarantees uniqueness, does it not?

Reply:
> If list indexes were fine as keys, then React wouldn't need to have a concept of key in the first place.

They are absolutely necessary when state is involved. And I tried to argue that we need keys whenever state and conditionally rendered components collide, even without arrays. But React does not warn about those, so we direct all our energy at arrays full of harmless stateless divs...

Reply:
> Using list indices guarantees uniqueness, does it not?

Yes, but that's not sufficient. You need to guarantee that "the content does not change if they key does not change", and you can't do that with list indexes. E.g. if you have [one, three, five], and insert a "two" at the second position, "two" would have the same key (2) as "three"  had before, and boom, you get a wrong render.

Reply:
> You need to guarantee that "the content does not change if they key does not change",

What is the content? Are you saying that a component with a key needs to stay completely static? No change in properties? That it needs to produce the same DOM elements as long as it has the same key?

Reply:
> Entries in a rendered array mustn't be replaced by a different component (whose properties or values don't change) with the same key, as then there will not be a re-render to actually update it.

That would surprise me. I'll try it out.

Reply:
We did a lot of cleanups related to this in the past, like commit 15db1e2b2ca02d7 or be6cd41ff61df7718 or 7c041b8f8574cf (that even has a bugzilla attached). This is clearly documented in https://react.dev/learn/rendering-lists#rules-of-keys with a big fat red warning, and we really should not break this in a library component. It may be acceptable in a top-level component where you absolutely know what you are doing, but really, why tempt it? The natural key for this case is `value`.

Reply:
These bugs are either about duplicated keys, or involve state. I think you are not willing to make the distinction between stateful and stateless components, and insist on applying the rules conservatively, as if state might get involved at any time and then we have a mess. Fair enough, I'll remove the internal key generation.

Reply:
> That would surprise me. I'll try it out.

Please join me in #21195 if you still want to convince me.

Reply:
> they are of course subject to their own re-renderings if their properties change

Property changes do not trigger re-renders. Re-renders trigger property changes.
State changes or React.render trigger re-renders.

Reply:
Let's just say it like that: there is very clear documentation what a key should do, we don't have control over what the caller puts into options, and there is zero reason to violate this rule here, so indeed I don't want to spend much time arguing for or thinking about why this *might* work.

Reply:
> Why do we need keys with this
> 
> ```
> const names = [];
> if (cond1)
>   names.push("A");
> if (cond2)
>   names.push("B");
> <div>
>   { names.map(x => <span>{c}</span> }
> </div>
> ```
> but not with this
> 
> ```
> <div>
>   { cond1 && <span>A</span> }
>   { cond2 && <span>B</span> }
> </div>
> ```

Because the latter is always two elements, either [ false, <span /> ], or [ <span />, false ], and that is enough to keep the state sorted. TIL!

https://react.dev/learn/preserving-and-resetting-state